store/src/java/com/zimbra/cs/ml/classifier/ClassifierRegistry.java

```diff
@@ -0,0 +1,197 @@
+                ZimbraLog.ml.warn("classifier id=%s has no corresponding entry in %s", cId, this.getClass().getSimpleName());
+            }
+        }
+        loaded = true;
+    }
+
+    public Map<String, Classifier<?>> getAllClassifiers() throws ServiceException {
+        if (!loaded) {
+            loadKnownClassifiers();
+        }
+        return idMap;
+    }
+
+    public boolean labelExists(String label) throws ServiceException {
+        if (!loaded) {
+            loadKnownClassifiers();
+        }
+        return labelMap.containsKey(label);
+    }
+
+    /**
+     * return known encoded classifiers
+     */
+    protected abstract String[] load() throws ServiceException;
+
+    /**
+     * Persist the encoded classifier string
+     */
+    protected abstract void save(String encodedClassifier) throws ServiceException;
+
+    /**
+     * Register a classifier
+     */
+    public void register(Classifier<?> classifier) throws ServiceException {
+        if (!loaded) {
+            loadKnownClassifiers();
+        }
+        if (labelExists(classifier.getLabel())) {
+            throw ServiceException.FAILURE("label '" + classifier.getLabel() + "' already exists", null);
+        }
+        idMap.put(classifier.getId(), classifier);
+        labelMap.put(classifier.getLabel(), classifier);
+        save(encode(classifier));
+    }
+
+    protected String encode(Classifier<?> classifier) {
+        Map<String, Object> map = new HashMap<>();
+        map.put(KEY_ID, classifier.getId());
+        map.put(KEY_TYPE, classifier.getType().name());
+        map.put(KEY_LABEL, classifier.getLabel());
+        map.put(KEY_FEATURES, classifier.getFeatureSet().getAllFeatureSpecs().stream().map(feature -> ((FeatureSpec<?>) feature).encode()).collect(Collectors.toList()));
+        map.put(KEY_DESCRIPTION, classifier.getDescription());
+        return BEncoding.encode(map);
+    }
+
+    protected Classifier<?> decode(String encoded) throws ServiceException {
+        Map<String, Object> map;
+        try {
+            map = BEncoding.decode(encoded);
+        } catch (BEncodingException e) {
+            throw ServiceException.FAILURE("unable to decode classifier", null);
+        }
+        if (!map.containsKey(KEY_TYPE)) {
+            throw ServiceException.FAILURE("no ClassifiableType value found during decoding classifier", null);
+
+        }
+        String typeStr = (String) map.get(KEY_TYPE);
+        ClassifiableType type = null;
+        try {
+            type = ClassifiableType.valueOf(typeStr);
+        } catch (IllegalArgumentException e) {
+            throw ServiceException.FAILURE("invalid ClassifiableType value encountered during decoding classifier: " + typeStr, null);
+        }
+        //this may seem unnecessary now, but it allows us to add classifiers for things other than messages in the future
+        switch (type) {
+        case MESSAGE:
+        default:
+            return decodeMessageClassifier(map);
```

Comment:
If the value of `type` isn't MESSAGE then I suspect this will make no sense.  Suggest treat anything other than MESSAGE as an error.

Reply:
Right, I think I filed that away as a "fix later" and then forgot about it. Thanks for noticing.